stl/inc/coroutine

```diff
@@ -0,0 +1,168 @@
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _COROUTINE_
+#define _COROUTINE_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _RESUMABLE_FUNCTIONS_SUPPORTED
+#pragma message("The contents of <coroutine> are not available with /await. Use <experimental/coroutine> instead.")
+#else // ^^^ /await ^^^ / vvv no /await vvv
+#ifndef __cpp_lib_coroutine
+#pragma message("The contents of <coroutine> are available only with C++20 or later.")
+#else // ^^^ __cpp_lib_coroutine not defined / __cpp_lib_coroutine defined vvv
+#ifndef _ALLOW_COROUTINE_ABI_MISMATCH
+#pragma detect_mismatch("_COROUTINE_ABI", "2")
+#endif // __ALLOW_COROUTINE_ABI_MISMATCH
+
+#include <compare>
+#include <type_traits>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+// STRUCT TEMPLATE coroutine_traits
+template <class _Ret, class = void>
+struct _Coroutine_traits {};
+
+template <class _Ret>
+struct _Coroutine_traits<_Ret, void_t<typename _Ret::promise_type>> {
+    using promise_type = typename _Ret::promise_type;
+};
+
+template <class _Ret, class...>
+struct coroutine_traits : _Coroutine_traits<_Ret> {};
+
+// STRUCT TEMPLATE coroutine_handle
+template <class = void>
+struct coroutine_handle;
+
+template <>
+struct coroutine_handle<void> {
+    constexpr coroutine_handle() noexcept = default;
+    constexpr coroutine_handle(nullptr_t) noexcept {}
+
+    coroutine_handle& operator=(nullptr_t) noexcept {
+        _Ptr = nullptr;
+        return *this;
+    }
+
+    _NODISCARD constexpr void* address() const noexcept {
+        return _Ptr;
+    }
+
+    _NODISCARD static constexpr coroutine_handle from_address(void* const _Addr) noexcept { // strengthened
+        coroutine_handle _Result;
+        _Result._Ptr = _Addr;
+        return _Result;
+    }
+
+    constexpr explicit operator bool() const noexcept {
+        return _Ptr != nullptr;
+    }
+
+    _NODISCARD bool done() const noexcept { // strengthened
+        return __builtin_coro_done(_Ptr);
+    }
+
+    void operator()() const noexcept { // strengthened
+        __builtin_coro_resume(_Ptr);
+    }
+
+    void resume() const noexcept { // strengthened
+        __builtin_coro_resume(_Ptr);
+    }
+
+    void destroy() const noexcept { // strengthened
+        __builtin_coro_destroy(_Ptr);
+    }
+
+protected:
+    void* _Ptr = nullptr;
+};
+
+template <class _Promise>
+struct coroutine_handle : coroutine_handle<> {
+    using coroutine_handle<>::coroutine_handle;
+
+    _NODISCARD static coroutine_handle from_promise(_Promise& _Prom) noexcept { // strengthened
+        const auto _PromPtr  = const_cast<void*>(static_cast<const volatile void*>(_STD addressof(_Prom)));
+        const auto _FramePtr = __builtin_coro_promise(_PromPtr, 0, true);

```

Comment:
Naming style: We generally use `_Ugly_snakes` for local variables (and `_UglyCamels` for template parameters), so these should be `_Prom_ptr` and `_Frame_ptr`.